### PR TITLE
Kernel: Mutex priority inheritance and synchronization improvements.

### LINKED ISSUE #2360: Wait Tree: empty "waiting for one of following objects"
waittree

### LINKED ISSUE #2395: Properly handle priority boosting when dealing with mutexes
The current implementation is not complete and seems incorrect. 
Some reversing of the 3DS kernel yielded the following pseudocode:

```
On first mutex lock:
    mutex->prio = thread->prio
    add(thread->held_mutexes, mutex)
    thread->prio = std::min(thread->base_prio, minPrio(thread->held_mutexes))

On mutex unlock:
    remove(thread->held_mutexes, mutex)
    thread->prio = std::min(thread->base_prio, minPrio(thread->held_mutexes))

On WaitSynch waited on the mutex
    mutex->prio = minPrio(mutex->waiting_threads)
    locking_thread->prio = std::min(locking_thread->base_prio, minPrio(locking_thread->held_mutexes))
```

I plan on implementing this along with some other synchronization fixes in an upcoming PR

### PR DESCRIPTION
**Pull request by Subv**
Not all syscalls should cause reschedules, the first commit attempts to remedy that, however, it still does not cover all cases.

See https://github.com/Subv/citra/issues/7 for progress on what's being done..

The other commits implement mutex priority inheritance as per the 3DS's kernel. The implementation is based on reverse engineering of the kernel.

Fixes #2395 
Fixes #2360

A commit by commit review of this should be easier.

### CODE REVIEW

File: src/core/hle/kernel/kernel.cpp [hunk 1]
```diff
@@ -39,21 +39,19 @@ SharedPtr<Thread> WaitObject::GetHighestPriorityReadyThread() {
                thread->status == THREADSTATUS_DEAD;
```
Comment (Subv) [inline src/core/hle/kernel/kernel.cpp]: Mutexes should update their priority when a thread is removed from their waitlist

File: src/core/hle/kernel/mutex.cpp [hunk 2]
```diff
@@ -6,26 +6,33 @@
...
- * @param mutex The mutex that some thread is waiting on
+ * Boost's a thread's priority to the best priority among the thread's held mutexes.
+ * This prevents priority inversion via priority inheritance.
  */
-static void ResumeWaitingThread(Mutex* mutex) {
-    // Reset mutex lock thread handle, nothing is waiting
-    mutex->lock_count = 0;
-    mutex->holding_thread = nullptr;
-    mutex->WakeupAllWaitingThreads();
+static void UpdateThreadPriority(Thread* thread) {
```
Comment (bunnei) [inline src/core/hle/kernel/mutex.cpp]: This should be a member of Thread and take the mutex's priority as an arg

File: src/core/hle/kernel/mutex.cpp [hunk 3]
```diff
@@ -40,52 +47,69 @@ SharedPtr<Mutex> Mutex::Create(bool initial_locked, std::string name) {
...
+        mutex->Acquire(GetCurrentThread());
 
     return mutex;
 }
 
-bool Mutex::ShouldWait() {
-    auto thread = GetCurrentThread();
-    bool wait = lock_count > 0 && holding_thread != thread;
+bool Mutex::ShouldWait(Thread* thread) const {
+    if (thread == holding_thread)
```
Comment (bunnei) [inline src/core/hle/kernel/mutex.cpp]: Compare GetObjectId() and pass in a `const Thread&`?

File: src/core/hle/kernel/mutex.cpp [hunk 4]
```diff
@@ -40,52 +47,69 @@ SharedPtr<Mutex> Mutex::Create(bool initial_locked, std::string name) {
...
-}
-
-void Mutex::Acquire() {
-    Acquire(GetCurrentThread());
+    return lock_count > 0;
 }
 
-void Mutex::Acquire(SharedPtr<Thread> thread) {
-    ASSERT_MSG(!ShouldWait(), "object unavailable!");
+void Mutex::Acquire(Thread* thread) {
```
Comment (bunnei) [inline src/core/hle/kernel/mutex.cpp]: Why did you change this to a dumb pointer?

File: src/core/hle/kernel/mutex.h [hunk 5]
```diff
@@ -35,17 +35,15 @@ class Mutex final : public WaitObject {
     }
 
     int lock_count;                   ///< Number of times the mutex has been acquired
+    u32 priority;                     ///< The priority of the mutex, used for priority inheritance.
```
Comment (bunnei) [inline src/core/hle/kernel/mutex.h]: s32?

File: src/core/hle/kernel/kernel.cpp [hunk 6]
```diff
@@ -34,15 +33,17 @@ void WaitObject::RemoveWaitingThread(Thread* thread) {
...
-    boost::range::remove_erase_if(waiting_threads, [](const SharedPtr<Thread>& thread) {
-        return thread->status == THREADSTATUS_RUNNING || thread->status == THREADSTATUS_READY ||
-               thread->status == THREADSTATUS_DEAD;
-    });
-
-    // TODO(Subv): This call should be performed inside the loop below to check if an object can be
-    // acquired by a particular thread. This is useful for things like recursive locking of Mutexes.
-    if (ShouldWait())
-        return nullptr;
+    auto to_remove = waiting_threads.end();
```
Comment (Subv) [inline src/core/hle/kernel/kernel.cpp]: Eeeeh, i'm open to suggestions on how to do this in a better way.
Essentially what i want is to call RemoveWaitingThread on each of the threads marked for removal in this loop.

Comment (Subv) [inline src/core/hle/kernel/kernel.cpp]: I suppose this piece of code can go back into ResumeFromWait() now that we have Thread::wait_objects_ordered

File: src/core/hle/kernel/thread.h [hunk 7]
```diff
@@ -178,14 +186,18 @@ class Thread final : public WaitObject {
...
     boost::container::flat_set<SharedPtr<Mutex>> held_mutexes;
 
+    /// Mutexes that this thread is currently waiting for.
+    boost::container::flat_set<SharedPtr<Mutex>> pending_mutexes;
+
     SharedPtr<Process> owner_process; ///< Process that owns this thread
 
     /// Objects that the thread is waiting on.
     /// This is only populated when the thread should wait for all the objects to become ready.
     std::vector<SharedPtr<WaitObject>> wait_objects;
```
Comment (Subv) [inline src/core/hle/kernel/thread.h]: We can get rid of this one, but that'll probably happen in another PR.

File: src/core/hle/kernel/mutex.h (hunk 5 above)
Comment (Subv) [inline src/core/hle/kernel/mutex.h]: I'm fairly certain you can't have negative priorities

File: src/core/hle/kernel/mutex.cpp [hunk 8]
```diff
@@ -90,6 +87,9 @@ void Mutex::RemoveWaitingThread(Thread* thread) {
 }
 
 void Mutex::UpdatePriority() {
+    if (!holding_thread)
```
Comment (Subv) [inline src/core/hle/kernel/mutex.cpp]: Mutex::Release -> Mutex::WakeupAllWaitingThreads -> Mutex::UpdatePriority with holding_thread = nullptr

File: src/core/hle/svc.cpp [hunk 9]
```diff
@@ -395,17 +390,14 @@ static ResultCode WaitSynchronizationN(s32* out, Kernel::Handle* handles, s32 ha
...
-
         // Add the thread to each of the objects' waiting threads.
         for (size_t i = 0; i < objects.size(); ++i) {
             Kernel::WaitObject* object = objects[i].get();
-            // Set the index of this object in the mapping of Objects -> index for this thread.
-            thread->wait_objects_index[object->GetObjectId()] = static_cast<int>(i);
             object->AddWaitingThread(thread);
         }
 
+        thread->wait_objects = std::move(objects);
```
Comment (yuriks) [inline src/core/hle/svc.cpp]: Looks like you can do this before the current `if` so you don't need to repeat it in both branches.

File: src/core/hle/svc.cpp [hunk 10]
```diff
@@ -277,6 +277,7 @@ static ResultCode WaitSynchronization1(Kernel::Handle handle, s64 nano_seconds)
         if (nano_seconds == 0)
             return ERR_SYNC_TIMEOUT;
 
+        thread->wait_objects = {object};
```
Comment (yuriks) [inline src/core/hle/svc.cpp]: Should you be clearing wait_objects if ShouldWait is false? (I'm guessing it doesn't matter.)

File: src/core/hle/kernel/kernel.cpp [hunk 11]
```diff
@@ -55,10 +55,16 @@ SharedPtr<Thread> WaitObject::GetHighestPriorityReadyThread() {
         if (ShouldWait(thread.get()))
             continue;
 
-        bool ready_to_run = std::none_of(thread->wait_objects.begin(), thread->wait_objects.end(),
+        // A thread is ready to run if it's either in THREADSTATUS_WAIT_SYNCH_ANY or
+        // in THREADSTATUS_WAIT_SYNCH_ALL and the rest of the objects it is waiting on are ready.
```
Comment (Subv) [inline src/core/hle/kernel/kernel.cpp]: The default is for the ANY case, there's an assert up there that ensures only these two cases can happen.

File: src/core/hle/svc.cpp (hunk 9 above)
Comment (Subv) [inline src/core/hle/svc.cpp]: Which if? the wait_all check? I'd have to add clearing code to both branches anyways if the thread did not have to wait

**Comment by yuriks** [comment]
FYI There's a formatting error in Travis.

**Pull request merged**